flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java

```diff
@@ -419,6 +435,14 @@ public void merge(W mergeResult,
 				registerCleanupTimer(window);
 			}
 		}
+
+		// side output input event if
+		// element not handled by any window
+		// late arriving tag has been set
+		// windowAssigner is event time and current timestamp + allowed lateness no less than element timestamp
+		if(isSkippedElement && lateDataOutputTag != null && isLate(element)) {
```

Comment:
@chenqin and @aljoscha I am starting to review the PR and I was wondering when is this new `isLate()` check needed? At least for the out-of-box window assigners, this seems to be a redundant check.